apollo-router/src/protocols/websocket.rs

```diff
@@ -414,11 +431,32 @@ where
                 },
                 None => Poll::Ready(None),
             },
-            Poll::Pending => Poll::Pending,
+            Poll::Pending => {
+                if let Some(heartbeat_interval) = this.heartbeat_interval {
+                    match heartbeat_interval.poll_tick(cx) {
+                        Poll::Ready(_) => send_heartbeat(this.stream, cx),
+                        Poll::Pending => (),
+                    };
+                }
+                Poll::Pending
+            }
         }
     }
 }
 
+fn send_heartbeat(mut stream: Pin<&mut impl Sink<ClientMessage>>, cx: &mut std::task::Context<'_>) {
+    if stream.as_mut().poll_flush(cx).map(|result| result.is_ok()) == Poll::Ready(true)
+        && stream.as_mut().poll_ready(cx).map(|result| result.is_ok()) == Poll::Ready(true)
```

Comment:
I like my original version with pattern matching: https://github.com/apollographql/router/pull/4802/commits/efcb2bfb10d242ab34cca5598e03df73eeb2d6b8#diff-b932e2ce79bcb30a9685d96419b458d13772deb85c850fe19741532cc21f0dbeL448

But Clippy forced me to rewrite it in this form.

Reply:
I was a little bit confused by the name of the variable. I think you should name it `sink`

Reply:
you right! will change

Reply:
Rewrote this part completely.